
{%include "base.html" %}
    <title> {% block head_title %} Results for FDA Assay|UVic MRMAssayDB{% endblock %} </title>


    {% load staticfiles %}
    <link href="https://cdn.datatables.net/1.10.16/css/jquery.dataTables.min.css" rel="stylesheet">
    <link href="{% static 'css/select.dataTables.min.css' %}" rel="stylesheet">
    <link href="{% static 'css/buttons.dataTables.min.css' %}" rel="stylesheet">    
    <link href="{% static 'css/bootstrap.min.css' %}" rel="stylesheet">
    <link href="{% static 'css/navbar-fixed-top.css' %}" rel="stylesheet">
    
    <!-- <script src="{% static 'js/jquery-1.12.0.min.js' %}"></script> -->
    <script src="https://code.jquery.com/jquery-1.12.4.js"></script>
    <script src="https://cdn.datatables.net/1.10.16/js/jquery.dataTables.min.js"></script>
    <script src="{% static 'js/dataTables.select.min.js' %}"></script>
    <script src="{% static 'js/dataTables.buttons.min.js' %}"></script>
    <script src="{% static 'js/buttons.flash.min.js' %}"></script>
    <script src="{% static 'js/vfs_fonts.js' %}"></script>
    <script src="{% static 'js/buttons.html5.min.js' %}"></script>
    <script src="{% static 'js/buttons.colVis.min.js' %}"></script>
    <script src="{% static 'js/loader.js' %}"></script>
    <script src="{% static 'js/canvas2svg.js' %}"></script>
    <script src="{% static 'js/jvenn.min.js' %}"></script>

    <style> 
        .my_button_space_class{
            margin-right :20px; 
        }
    </style>
    <style type="text/css">
        div.dataTables_wrapper {
            margin-bottom: 3em;
        }
    </style>
    <script type="text/javascript">
        $(document).ready(function() {
        // Setup - add a text input to each footer cell
        $('#dataTables-wrkld tfoot th').filter(':eq(1),:eq(2),:eq(3),:eq(4),:eq(5),:eq(6),:eq(7),:eq(8),:eq(9),:eq(10),:eq(11),:eq(12),:eq(13),:eq(14),:eq(15),:eq(16),:eq(17),:eq(18),:eq(19),:eq(20),:eq(21),:eq(22),:eq(23),:eq(24),:eq(25),:eq(26),:eq(27),:eq(28),:eq(29)').each(function() {
            var title = $(this).text();
            $(this).html('<input type="text" placeholder="Filter ' + title + '" />');
        });

        var colnames={{colname|safe}};
        var tempcol=[];
        for (ci = 0; ci < colnames.length;ci++) {
            tempcol.push({
                "data":colnames[ci]
            });

        };
        var table = $('#dataTables-wrkld').dataTable({
            processing: true,
            serverSide: false,
            orderCellsTop: true,
            fixedHeader: true,
            ajax: {
                url: {{filename|safe}},
                dataType: 'json',
                type: 'GET',
                data: function(args) {
                    return {
                        "args": JSON.stringify(args)
                    };
                },
                dataSrc: "",
            },
            columns: tempcol,
            scrollX:true,
            scrollY:'650px',
            scrollCollapse:true,
            dom: 'Blfrtip',
            autoWidth:false,
            buttons:[
            {
                extend: 'colvis',
                columns: ':not(.permanent)'
            },
            {
                extend:'csv',
                filename: 'resultsmrmassaydb',
                text:'Download all',
                customize: function (csv) {
                    var csvRows = csv.split('\n');
                    var csvCols=csvRows[0].split(',');
                    csvCols[4] = csvCols[4].replace(csvCols[4], '"Subcellular localization"');
                    csvCols[10] = csvCols[10].replace(csvCols[10], '"PeptideTracker Transitions"');
                    csvCols[12] = csvCols[12].replace(csvCols[12], '"PASSEL Transitions"');
                    csvCols[14] = csvCols[14].replace(csvCols[14], '"SRMAtlas Transitions"');
                    csvCols[16] = csvCols[16].replace(csvCols[16], '"CPTAC Transitions"');
                    csvCols[18] = csvCols[18].replace(csvCols[18], '"PanoramaWeb Transitions"');
                    csvCols[21] = csvCols[21].replace(csvCols[21], '"Pathway Name"');
                    csvCols[22] = csvCols[22].replace(csvCols[22], '"UniProt DiseaseData"');
                    csvCols[23] = csvCols[23].replace(csvCols[23], '"DisGen DiseaseData"');
                    csvCols[24] = csvCols[24].replace(csvCols[24], '"Go Term Name"');
                    csvCols[25] = csvCols[25].replace(csvCols[25], '"Drug Association"');
                    csvCols[27] = csvCols[27].replace(csvCols[27], '"Concentration"');
                    csvRows[0] = csvCols.join(',');
                    return csvRows.join('\n');
                },
                exportOptions:{
                    columns:  [1,2,3,4,38,6,7,8,9,10,31,12,32,14,33,16,34,18,35,20,21,37,45,46,36,47,28,39]
                }
            },
            {
                extend:'csv',
                filename:'resultsmrmassaydbselected',
                text:'Download selected',
                customize: function (csv) {
                    var csvRows = csv.split('\n');
                    var csvCols=csvRows[0].split(',');
                    csvCols[4] = csvCols[4].replace(csvCols[4], '"Subcellular localization"');
                    csvCols[10] = csvCols[10].replace(csvCols[10], '"PeptideTracker Transitions"');
                    csvCols[12] = csvCols[12].replace(csvCols[12], '"PASSEL Transitions"');
                    csvCols[14] = csvCols[14].replace(csvCols[14], '"SRMAtlas Transitions"');
                    csvCols[16] = csvCols[16].replace(csvCols[16], '"CPTAC Transitions"');
                    csvCols[18] = csvCols[18].replace(csvCols[18], '"PanoramaWeb Transitions"');
                    csvCols[21] = csvCols[21].replace(csvCols[21], '"Pathway Name"');
                    csvCols[22] = csvCols[22].replace(csvCols[22], '"UniProt DiseaseData"');
                    csvCols[23] = csvCols[23].replace(csvCols[23], '"DisGen DiseaseData"');
                    csvCols[24] = csvCols[24].replace(csvCols[24], '"Go Term Name"');
                    csvCols[25] = csvCols[25].replace(csvCols[25], '"Drug Association"');
                    csvCols[27] = csvCols[27].replace(csvCols[27], '"Concentration"');
                    csvRows[0] = csvCols.join(',');
                    return csvRows.join('\n');
                },
                exportOptions:
                {
                    columns: [1,2,3,4,38,6,7,8,9,10,31,12,32,14,33,16,34,18,35,20,21,37,45,46,36,47,28,39],
                    modifier: {selected: true}
                }
            }
            ],
            columnDefs:[{
                orderable: false, 
                className: 'select-checkbox',
                searchable: false,
                targets:0
            },
            {
              targets:1,
                    render: function (data, type, row) {
                        return '<a target="_blank" href="https://www.uniprot.org/uniprot/' + row["UniProtKB Accession"].trim() + '">' + row["UniProtKB Accession"].trim() + '</a>';

                    }
            },
            {
              targets:2,
                    render: function (data, type, row) {
                        if ((row["Protein"].trim()).length > 0){
                          return '<a target="_blank"  href="{% url "mutptmdom" %}?Uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() +'">' + row["Protein"].trim() + '</a>';
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:3,
                    render: function (data, type, row) {
                        if ((row["Gene"].trim()).length > 0){
                            return row["Gene"].trim();
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:4,
                    render: function (data, type, row) {
                        if ((row["Organism"].trim()).length > 0){
                            return (row["Organism"].trim()).split(';')[0];
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:5,
                    render: function (data, type, row) {
                        if ((row["SubCellular"].trim()).length > 0){
                            return (row["SubCellular"].trim()).split('|').join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:6,
                    render: function (data, type, row) {
                        return '<a target="_blank"  href="{% url "mutptmdom" %}?Uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() +'">' + row["Peptide Sequence"].trim() + '</a>';

                    }
            },
            {
              targets:7,
                    render: function (data, type, row) {
                        return row["Modified Peptide Sequence"].trim();
                    }
            },
            {
              targets:8,
                    render: function (data, type, row) {
                        if ((row["Unique in protein"].trim()).length > 0 & (row["Unique in protein"].trim()) == "Yes"){
                            return (row["Unique in protein"].trim());
                        } else {
                            return 'Not unique';
                        }
                    }
            },
            {
              targets:9,
                    render: function (data, type, row) {
                        if ((row["Present in isoforms"].trim()).length > 0 & (row["Present in isoforms"].trim()) != "No"){
                            var array =(row["Present in isoforms"].trim()).split(',');
                            array.forEach(function(element, index) {
                                array[index] = '<a target="_blank"  href="https://www.uniprot.org/uniprot/' + element+'.fasta">'+element+ '</a>';
                            });
                            return array.join("<br>");
                        } else {
                            return 'No';
                        }
                    }
            },
            {
              targets:10,
                    render: function (data, type, row) {
                        if ((row["PeptideTracker ID"].trim()).length > 0 & row["PeptideTracker ID"].trim() != 'NA'){
                            return (row["PeptideTracker URL"].trim()).split('|').join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:11,
                    render: function (data, type, row) {
                        if ((row["PeptideTracker Transition"].trim()).length > 0 & (row["PeptideTracker ID"].trim()).length > 0 & (row["PeptideTracker Transition"].trim()) != "NA"){
                            var tempURL='<a target="_blank"  href="{% url "transition" %}?uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() + '&modpepseq='+row["Modified Peptide Sequence"].trim() +'">' + '....more' + '</a>';
                            var temprowdata = [row["PeptideTracker TransView"].trim(),tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:12,
                    render: function (data, type, row) {
                        if ((row["Passel ID"].trim()).length > 0& row["Passel ID"].trim() != 'NA'){
                            return (row["Passel URL"].trim()).split('|').join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:13,
                    render: function (data, type, row) {
                        if ((row["Passel Transition"].trim()).length > 0 & (row["Passel ID"].trim()).length > 0 & (row["Passel Transition"].trim()) != "NA"){
                            var tempURL='<a target="_blank"  href="{% url "transition" %}?uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() + '&modpepseq='+row["Modified Peptide Sequence"].trim() +'">' + '....more' + '</a>';
                            var temprowdata = [row["Passel TransView"].trim(),tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:14,
                    render: function (data, type, row) {
                        if ((row["SRMAtlas ID"].trim()).length > 0 & row["SRMAtlas ID"].trim() != 'NA'){
                            return (row["SRMAtlas URL"].trim()).split('|').join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:15,
                    render: function (data, type, row) {
                        if ((row["SRMAtlas Transition"].trim()).length > 0 & (row["SRMAtlas ID"].trim()).length > 0 & (row["SRMAtlas Transition"].trim()) != "NA"){
                            var tempURL='<a target="_blank"  href="{% url "transition" %}?uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() + '&modpepseq='+row["Modified Peptide Sequence"].trim() +'">' + '....more' + '</a>';
                            var temprowdata = [row["SRMAtlas TransView"].trim(),tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:16,
                    render: function (data, type, row) {
                        if ((row["Cptac ID"].trim()).length > 0 & row["Cptac ID"].trim() != 'NA'){
                            return (row["CPTAC URL"].trim()).split('|').join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:17,
                    render: function (data, type, row) {
                        if ((row["CPTAC Transitions"].trim()).length > 0 & (row["Cptac ID"].trim()).length > 0 & (row["CPTAC Transitions"].trim()) != "NA"){
                            var tempURL='<a target="_blank"  href="{% url "transition" %}?uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() + '&modpepseq='+row["Modified Peptide Sequence"].trim() +'">' + '....more' + '</a>';
                            var temprowdata = [row["CPTAC TransView"].trim(),tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:18,
                    render: function (data, type, row) {
                        if ((row["Panoramaweb ID"].trim()).length >0 & row["Panoramaweb ID"].trim() != 'NA'){
                            return (row["Panoramaweb URL"].trim()).split('|').join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:19,
                    render: function (data, type, row) {
                        if ((row["Panoramaweb Transition"].trim()).length > 0 & (row["Panoramaweb ID"].trim()).length > 0 & (row["Panoramaweb Transition"].trim()) != "NA"){
                            var tempURL='<a target="_blank"  href="{% url "transition" %}?uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() + '&modpepseq='+row["Modified Peptide Sequence"].trim() +'">' + '....more' + '</a>';
                            var temprowdata = [row["Panoramaweb TransView"].trim(),tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:20,
                    render: function (data, type, row) {
                        return row["Peptide Occurrence"].trim();
                    }
            },
            {
              targets:21,
                    render: function (data, type, row) {
                        if ((row["Best Transition"].trim()).length > 0){
                            var tempURL='<a target="_blank"  href="{% url "transition" %}?uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() + '&modpepseq='+row["Modified Peptide Sequence"].trim() +'">' + '....more' + '</a>';
                            var temprowdata = (row["Best Transition"].trim()).split('|');
                            temprowdata.push(tempURL)
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:22,
                    render: function (data, type, row) {
                        if ((row["Pathway Name"].trim()).length > 0 & (row["Pathway Name"].trim()) != "NA"){
                            var tempURL='<a target="_blank" href="{% url "pathway" %}?Uniprotkb='+ row["UniProtKB Accession"].trim() +'&organismid='+row["Organism ID"].trim()+'#KeggDataTable">' + '...more' + '</a>';
                            var tempdata=(row["Pathway Name"].trim()).split("|").slice(0, 3).join("<br>");
                            var temprowdata=[tempdata,tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }
                        
                    }
            },
            {
              targets:23,
                    render: function (data, type, row) {
                        var tempURL='<a target="_blank" href="{% url "ppi" %}?Uniprotkb='+ row["UniProtKB Accession"].trim() +'&organismid='+row["Organism ID"].trim()+'#string">' + 'View' + '</a>';
                        return tempURL;
                    }
            },
            {
              targets:24,
                    render: function (data, type, row) {
                        if ((row["Disease Name"].trim()).length > 0 & (row["Disease Name"].trim()) != "NA"){
                            var tempURL='<a target="_blank"  href="{% url "disease" %}?Uniprotkb=' + row["UniProtKB Accession"].trim() +'">' + '....more' + '</a>';
                            var tempdata=(row["Disease Name"].trim()).split("|").slice(0, 3).join("<br>");
                            var temprowdata=[tempdata,tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }
                        
                    }
            },
            {
              targets:25,
                    render: function (data, type, row) {
                        var tempURL='<a target="_blank"  href="{% url "mutptmdom" %}?Uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() +'">' + 'View' + '</a>';
                        return tempURL;
                        
                    }
            },
            {
              targets:26,
                    render: function (data, type, row) {
                        if ((row["Go Name"].trim()).length > 0 & (row["Go Name"].trim()) != "NA"){
                            var tempURL='<a target="_blank" href="{% url "goterm" %}?Uniprotkb='+ row["UniProtKB Accession"].trim() +'">' + '....more' + '</a>';
                            var tempdata=(row["Go Name"].trim()).split("|").slice(0, 3).join("<br>");
                            var temprowdata=[tempdata,tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return 'NA';
                        }

                    }
            },
            {
              targets:27,
                    render: function (data, type, row) {
                        if ((row["Drug Bank"].trim()).length > 0 & (row["Drug Bank"].trim()) != "NA"){
                            var tempURL='<a target="_blank"  href="{% url "drug" %}?Uniprotkb=' + row["UniProtKB Accession"].trim() +'">' + '....more' + '</a>';
                            var tempdata=(row["Drug Bank"].trim()).split('|').slice(0, 3).join("<br>");
                            var temprowdata=[tempdata,tempURL];
                            return temprowdata.join("<br>");
                        }else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:28,
                    render: function (data, type, row) {
                        return row["Assays for FDA approved Marker"].trim()
                    }
            },
            {
              targets:29,
                    render: function (data, type, row) {
                        if ((row["Assays for FDA approved Marker"].trim()).length > 0 & (row["Assays for FDA approved Marker"].trim()) != "No" & (row["Mean Concentration"].trim()) != "No" & (row["PeptideTracker ID"].trim()).length > 0 & row["PeptideTracker ID"].trim() != 'NA'){
                            var tempURL='<a target="_blank"  href="{% url "concentration" %}?uniprotkb=' + row["UniProtKB Accession"].trim() +'&pepseq='+row["Peptide Sequence"].trim() + '&peptrackid='+row["PeptideTracker ID"].trim() +'">' + '....more' + '</a>';
                            var temprowdata = [row["Mean Concentration"].trim(),tempURL];
                            return temprowdata.join("<br>");
                        } else {
                            return (row["Mean Concentration"].trim());
                        }
                    }
            },
            {
              targets:31,
                    render: function (data, type, row) {
                        if ((row["PeptideTracker Transition"].trim()).length > 0 & (row["PeptideTracker ID"].trim()).length > 0){
                            return row["PeptideTracker Transition"].trim();
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:32,
                    render: function (data, type, row) {
                        if ((row["Passel Transition"].trim()).length > 0 & (row["Passel ID"].trim()).length > 0){
                            return row["Passel Transition"].trim();
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:33,
                    render: function (data, type, row) {
                        if ((row["SRMAtlas Transition"].trim()).length > 0 & (row["SRMAtlas ID"].trim()).length > 0){
                            return row["SRMAtlas Transition"].trim();
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:34,
                    render: function (data, type, row) {
                        if ((row["CPTAC Transitions"].trim()).length > 0 & (row["Cptac ID"].trim()).length > 0){
                            return row["CPTAC Transitions"].trim();
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:35,
                    render: function (data, type, row) {
                        if ((row["Panoramaweb Transition"].trim()).length > 0 & (row["Panoramaweb ID"].trim()).length > 0){
                            return row["Panoramaweb Transition"].trim();
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:36,
                    render: function (data, type, row) {
                        if ((row["Go Name"].trim()).length > 0){
                            return row["Go Name"].trim();
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:37,
                    render: function (data, type, row) {
                        if ((row["Pathway Name"].trim()).length > 0){
                            return row["Pathway Name"].trim();
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:38,
                    render: function (data, type, row) {
                        if ((row["SubCellular"].trim()).length > 0){
                            return (row["SubCellular"].trim());
                        } else {
                            return 'NA';
                        }
                    }
            },
            {
              targets:39,
                    render: function (data, type, row) {
                        return row["Concentration"].trim()
                    }
            },
            {
              targets:47,
              className: 'permanent',
                    render: function (data, type, row) {
                        return row["Drug Bank"].trim()
                    }
            },
            {
                targets: [10,11,12,13,14,15,16,17,18,19,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47],
                visible: false,
                searchable: true
             } 

            ],
            select:{
                style:'multi',
                selector: 'td:first-child'
            },
            order:[[ 1, 'asc' ]],

            initComplete: function() {
                var api = this.api();
                // Apply the search
                var filterColumns = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29];
                api.columns(filterColumns).every(function() {
                    var that = this;
                    $('input', this.footer() ).on('keyup change', function() {
                        if (that.search() !== this.value) {
                            that
                            .search(this.value)
                            .draw();
                        }
                    });
                });
            }

        });
    });
    </script>
    <script type="text/javascript">
    $(document).ready(function() {
        $('table.summarystat').DataTable();
    } );
    </script>
    <style type="text/css">
        .cellContainer {
          width: 25%;
          float: left;
        }
    </style>

{% block content %}
    <div class="container-fluid">
            {% if foundHits > 0 %}
            <p class ='text-align-center'>You searched for: <strong>FDA Assay</strong></p>
                <style type="text/css">
                    tfoot {
                        display: table-header-group;
                    }
                </style>
                <br>
                  <div align="justify">
                    <div id="gallery-text">
                      <div class="centerBlock">
                            <p>
                                The proteomics community possesses a large and increasing amount of knowledge on precise and reproducible targeted assays, but these assays are not yet utilized to
                                their full potential, particularly for clinical applications. The US Food and Drug Administration (FDA) has listed 1140 approved clinical tests, 208 of which are
                                associated with 253 proteins. Using MRMAssayDB, an integrative resource on targeted proteomics assays, we have data-mined all of the available MRM assays to determine
                                which ones can be used for these FDA-approved markers. Because MRMAssayDB links all assays to DrugBank and other knowledge bases, and to protein-protein interactions,
                                it enables straight-forward design of multiplexed MRM assays for studying the pathways and interactions that involve these FDA-approved protein biomarkers.

                            </p>
                      </div>
                    </div>
                  </div>
                <br>
                <table  id="dataTables-wrkld" class="display cell-border" cellspacing="0" width="100%">
                    <thead>
                        <tr>
                            <th class="permanent"></th>
                            <th>UniProtKB Accession</th>
                            <th>Protein</th>
                            <th>Gene</th>
                            <th>Organism</th>
                            <th>Subcellular localization</th>
                            <th>Stripped Peptide Sequence</th>
                            <th>Modified Peptide Sequence</th>
                            <th>Peptide unique in proteome</th>
                            <th>Peptide present in isoforms</th>
                            <th>PeptideTracker ID</th>
                            <th>PeptideTracker Transitions</th>
                            <th>PASSEL ID</th>
                            <th>PASSEL Transitions</th>
                            <th>SRMAtlas ID</th>
                            <th>SRMAtlas Transitions</th>
                            <th>CPTAC ID</th>
                            <th>CPTAC Transitions</th>
                            <th>PanoramaWeb ID</th>
                            <th>PanoramaWeb Transitions</th>
                            <th>Number of repositories</th>
                            <th>Best Transitions</th>
                            <th>Biological Pathway</th>
                            <th>Protein–protein interaction</th>
                            <th>Disease Association</th>
                            <th>3D structure</th>
                            <th>GO Terms</th>
                            <th>Drug Association</th>
                            <th>Protein approved as FDA Marker</th>
                            <th>Mean Concentration</th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>
                            <th class="permanent"></th>                            
                        </tr>

                    </thead>
                    <tfoot>
                        <tr>
                            <th></th>
                            <th>UniProtKB Accession</th>
                            <th>Protein</th>
                            <th>Gene</th>
                            <th>Organism</th>
                            <th>Subcellular localization</th>
                            <th>Stripped Peptide Sequence</th>
                            <th>Modified Peptide Sequence</th>
                            <th>Peptide unique in proteome</th>
                            <th>Peptide present in isoforms</th>
                            <th>PeptideTracker ID</th>
                            <th>PeptideTracker Transitions</th>
                            <th>PASSEL ID</th>
                            <th>PASSEL Transitions</th>
                            <th>SRMAtlas ID</th>
                            <th>SRMAtlas Transitions</th>
                            <th>CPTAC ID</th>
                            <th>CPTAC Transitions</th>
                            <th>PanoramaWeb ID</th>
                            <th>PanoramaWeb Transitions</th>
                            <th>Number of repositories</th>
                            <th>Best Transitions</th>
                            <th>Biological Pathway</th>
                            <th>Protein–protein interaction</th>
                            <th>Disease Association</th>
                            <th>3D structure</th>
                            <th>GO Terms</th>
                            <th>Drug Association</th>
                            <th>Protein approved as FDA Marker</th>
                            <th>Mean Concentration</th>
                        </tr>
                        </tr>
                    </tfoot>
                </table>
                {% if foundHits > 10000 %}
                 (<font color="red">WARNING:</font> The maximum number of entries was reached. Result entries in the Web interface are limited to 10000. Please make the search more specific using multiple filters in the advanced search functionality or use the <a target="_blank" href="http://mrmassaydb.proteincentre.com/help/#rest"> REST API service to download all results</a>.)
                {% endif %}
            {% else %}
                <p>No Peptide matched with your search criteria.</p>
            {% endif %}
    </div>
{% endblock %}
{% block jumbotron %}
        <div class="container-fluid">
            <div class="jumbotron">
                {% if specieslist|length > 0 %}
                <p class ='text-align-center'>Summary of results</strong></p>
                    </br>
                    <div class="container">
                        <div style="width: 100%;">
                            <div style="width: 45%; float: left; margin: 5px;">
                                <caption><b>The number of peptides entries in MRMAssayDB, sorted by organism</b></caption>
                                <br>
                                <table id="" border="1" class="summarystat"  cellspacing="0" width="100%">
                                    <thead>
                                        <tr style="text-align: center;">
                                            <td align="center"><b>Species Name</b></td>
                                            <td align="center"><b>Number of assays</b></td>
                                            <td align="center"><b>Number of covered proteins</b></td>
                                        </tr>
                                    </thead>
                                    <tbody>
                                        {% for items in specieslist %}
                                            <tr>
                                                <td align="center">{{ items.0 }}</td>
                                                <td align="center">{{ items.2 }}</td>
                                                <td align="center">{{ items.1 }}</td>
                                            </tr>
                                        {% endfor %}
                                        <tr>
                                            <td align="center"><b>Total</b></td>
                                            <td align="center">{{ totallist.1 }}</td>
                                            <td align="center"> {{ totallist.0 }}</td>
                                        </tr>
                                    </tbody>
                                </table>
                            </div>
                            <div  style="width: 45%; float: right; margin: 5px;">
                                {% if unqisostat|length > 0 %}
                                    <caption><b>Uniqueness and presence in isofroms</b></caption>
                                    <br>
                                    <table id="" border="1" class="summarystat"  cellspacing="0" width="100%">
                                         
                                        <thead>
                                            <tr style="text-align: center;">
                                                <td align="center"><b>Database name</b></td>
                                                <td align="center"><b>Peptide uniqueness in proteome(%)</b></td>
                                                <td align="center"><b>Peptide is specific to one protein isoform(%)</b></td>
                                            </tr>
                                        </thead>
                                        <tbody>
                                            {% for items in unqisostat %}

                                                <tr>
                                                    <td align="center">
                                                        {% if  items.0|upper == "CPTAC" %}
                                                            {{ items.0|upper }}
                                                        {% elif  items.0|upper == "PASSEL" %}
                                                            {{ items.0|upper }}
                                                        {% elif  items.0|upper == "PANORAMAWEB" %}
                                                            PanoramaWeb
                                                        {% else %}
                                                            {{ items.0 }}
                                                        {% endif %}
                                                    </td>
                                                    <td align="center"> {{ items.1 }}</td>
                                                    <td align="center"> {{ items.2 }}</td>
                                                </tr>
                                            {% endfor %}
                                        </tbody>
                                    </table>
                                {% endif %}
                            </div>
                        </div>
                    </div>
                    <div class="container">
                        <div style="width: 100%;">
                            <div style="width: 50%; float: left; margin: 5px;">
                                {% if subcell|length > 0 %}
                                    <caption><b>The number of peptides entries in MRMAssayDB, sorted by sub cellular location (Top50)</b></caption>
                                    <br>
                                    <table id="" border="1" class="summarystat"  cellspacing="0" width="100%">
                                        <thead>
                                            <tr style="text-align: center;">
                                                <td align="center"><b>Sub cellular location</b></td>
                                                <td align="center"><b>Number of unique entries(Proteins)</b></td>
                                            </tr>
                                        </thead>
                                        <tbody>
                                            {% for key, value in subcell.items %}

                                                <tr>
                                                    <td align="center">{{ key }}</td>
                                                    <td align="center"> {{ value }}</td>
                                                </tr>
                                            {% endfor %}
                                        </tbody>
                                    </table>
                                {% endif %}
                            </div>
                            <div  style="width: 45%; float: right; margin: 5px;">
                                {% if updatedgodic|length > 0 %}
                                   <caption><b>The number of peptides entries in MRMAssayDB, sorted by GO term name (Top50)</b></caption>
                                   <br>
                                    <table id="" border="1" class="summarystat"  cellspacing="0" width="100%">
                                        <thead>
                                            <tr style="text-align: center;">
                                                <td align="center"><b>GO Term Name</b></td>
                                                <td align="center"><b>Number of unique entries(Proteins)</b></td>
                                            </tr>
                                        </thead>
                                        <tbody>
                                            {% for key, value in updatedgodic.items %}

                                                <tr>
                                                    <td align="center">{{ key }}</td>
                                                    <td align="center"> {{ value }}</td>
                                                </tr>
                                            {% endfor %}
                                        </tbody>
                                    </table>
                                {% endif %}
                            </div>
                        </div>
                    </div>
                <div class="row">
                    <div class="container">
                        {% if pathwaychart|length > 0 %}
                            </br>
                                <p ><b>Pathway coverage (Top 10)</b></p>
                                <div id="chart_div_pathway" style="width: 100%; height: 100%;" ></div>
                        {% endif %}
                        </br>
                        <div id="gallery-text">
                            <p ><b>Distribution of proteins and peptides among different databases.</b></p>
                            <div class="container-fluid">
                                <div class="row">
                                    <div align="justify">
                                        <div id="gallery-text">
                                            <div class="col-xs-6">
                                                <div class="centerBlock">
                                                    <h4><b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Unique Proteins</b></h4>
                                                    <br><br>
                                                    <div id="vennprot" style="width:auto height:auto";></div>
                                                </div>
                                            </div>
                                            <div class="col-xs-6">
                                                <div class="centerBlock">
                                                    <h4><b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Unique Peptides</b></h4>
                                                    <br><br>
                                                    <div id="vennpep" style="width:auto height:auto";></div>
                                                </div>
                                            </div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
                {% endif %}
            </div>
        </div>

        <script type="text/javascript">
            google.charts.load('current', {packages: ['corechart', 'bar']});
            google.charts.setOnLoadCallback(drawPathwayChart);

            function drawPathwayChart() {
                 var tempdatapathway={{pathwaychart|safe}};
                 var data = google.visualization.arrayToDataTable(tempdatapathway);
                 var options = {
                    'height':400,
                    bar: {groupWidth: "100%"},
                    legend: {position: 'none'},
                    hAxis: {
                      title: 'No. of assays',
                       minValue: 0,
                        textStyle : {
                            fontSize: 16 // or the number you want
                        },
                    },
                    axes: {
                        x: {
                          0: { side: 'top'}
                        }
                    },
                    vAxis: {
                      title: 'Pathway Name',
                        textStyle : {
                            fontSize: 16 // or the number you want
                        },
                    },
                    bars: 'horizontal'
                  };
                  var chart = new google.charts.Bar(document.getElementById('chart_div_pathway'));
                  chart.draw(data, google.charts.Bar.convertOptions(options));
                }
        </script>

        <style type="text/css">
            .centerBlock {
                display: table;
                margin: auto;
            }
        </style>
        <script type="text/javascript" charset="utf-8">
            var pepsets = {{ jvennpep|safe }};
            var mrmdbname={{ jvennmrmdb|safe }};
            $(document).ready(function(){
                    $('#vennpep').jvenn({
                        fontSize: "12px",
                        series: [{name:mrmdbname,values:pepsets}],
                        disableClick: true,
                        exporting: false,
                        displaySwitch: true,
                         shortNumber: false,
                        colors: ["rgb(0,102,0)","rgb(90,155,212)","rgb(241,90,96)","rgb(255,117,0)","rgb(255,117,0)"]

                    });
            });
        </script>
        <script type="text/javascript" charset="utf-8">
            var protsets = JSON.parse('{{ jvennprot|safe }}');
            var mrmdbname=JSON.parse('{{ jvennmrmdb|safe }}');
            $(document).ready(function(){
                    $('#vennprot').jvenn({
                        fontSize: "12px",
                        series: [{name:mrmdbname,values:protsets}],
                        disableClick: true,
                        exporting: false,
                        displaySwitch: true,
                         shortNumber: false,
                        colors: ["rgb(0,102,0)","rgb(90,155,212)","rgb(241,90,96)","rgb(255,117,0)","rgb(255,117,0)"]

                    });
            });
        </script>
{% endblock %}
